k2/csrc/ragged_test.cu

```diff
@@ -1430,53 +1391,148 @@ TEST(RaggedShapeOpsTest, TestRenumber) {
-    // so it returns
-    //         3 4 7 5 6 0 1 2   Note that it is stable
-    CheckArrayData(order, {3, 4, 7, 5, 6, 0, 1, 2});
+    Array1<int32_t> order = GetTransposeReordering(ragged, 6);
+    CheckArrayData(
+        order, {5, 6, 9, 0, 1, 8, 10, 11, 7, 2, 3, 4, 12, 16, 13, 14, 15, 17});
+    EXPECT_TRUE(context->IsCompatible(*order.Context()));
+  }
+}
+
+TEST(GetTransposeReordering, WithDuplicatesThreeAxes) {
+  //       col0   col1   col2    col3      col4      col5
+  // row0         a0,a1         b2,b3,b4
+  // row1  c5,c6          d7
+  // row2         e8
+  // row3   f9   g10,g11         h12
+  // row4                                i13,i14,i15
+  // row5                                 j16         k17
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    Array1<int32_t> col_indexes(
+        context, std::vector<int32_t>{1, 1, 3, 3, 3, 0, 0, 2, 1, 0, 1, 1, 3, 4,
+                                      4, 4, 4, 5});
+    Array1<int32_t> row_splits1(context, std::vector<int32_t>{0, 4, 6});
+    Array1<int32_t> row_splits2(context,
+                                std::vector<int32_t>{0, 5, 8, 9, 13, 16, 18});
+    RaggedShape shape =
+        RaggedShape3(&row_splits1, nullptr, -1, &row_splits2, nullptr, -1);
+    Ragged<int32_t> ragged(shape, col_indexes);
+    Array1<int32_t> order = GetTransposeReordering(ragged, 6);
+    CheckArrayData(
+        order, {5, 6, 9, 0, 1, 8, 10, 11, 7, 2, 3, 4, 12, 13, 14, 15, 16, 17});
     EXPECT_TRUE(context->IsCompatible(*order.Context()));
   }
 }
 
+TEST(ChangeSublistSize, TwoAxes) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    Array1<int32_t> row_splits1(context, std::vector<int32_t>{0, 2, 5});
+    RaggedShape src = RaggedShape2(&row_splits1, nullptr, -1);
+
+    int32_t size_delta = 2;
+    RaggedShape dst = ChangeSublistSize(src, size_delta);
+    CheckArrayData(dst.RowSplits(1), std::vector<int32_t>{0, 4, 9});
+
+    size_delta = -2;
+    dst = ChangeSublistSize(src, size_delta);
+    CheckArrayData(dst.RowSplits(1), std::vector<int32_t>{0, 0, 1});
+
+    size_delta = 0;
+    dst = ChangeSublistSize(src, size_delta);
+    CheckArrayData(dst.RowSplits(1), std::vector<int32_t>{0, 2, 5});
+  }
+}
+
+TEST(ChangeSublistSize, ThreeAxes) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    /*
+     [
+       [ [x, x, x], [x, x] ]
+       [ [x], [x, x], [x, x, x] ]
+     ]
+     */
+    Array1<int32_t> row_splits1(context, std::vector<int32_t>{0, 2, 5});
+    Array1<int32_t> row_splits2(context,
+                                std::vector<int32_t>{0, 3, 5, 6, 8, 11});
+    RaggedShape src =
+        RaggedShape3(&row_splits1, nullptr, -1, &row_splits2, nullptr, -1);
+
+    int32_t size_delta = 2;
+    RaggedShape dst = ChangeSublistSize(src, size_delta);
+    CheckArrayData(dst.RowSplits(2), std::vector<int32_t>{0, 5, 9, 12, 16, 21});
+
+    // it is an error to use -2 here
+    // because the state (state_idx01 == 2) has only 1 entry
```

Comment:
The program will crash if use `-2` here? I thought the implementation was supposed to use something like `max(0, size + delta)` to handle such case

Reply:
The current implementation
https://github.com/k2-fsa/k2/blob/c3306ca4cdc5528f32ae85e521c412a3ee09e918/k2/csrc/ragged_ops.cu#L954

assumes that no entry will have a negative size after adding `size_delta`.

https://github.com/k2-fsa/k2/blob/c3306ca4cdc5528f32ae85e521c412a3ee09e918/k2/csrc/ragged_ops.h#L156-L157